### ISSUE TITLE
Running Aria2 on a separate host

### ISSUE THREAD

**Issue by adocampo** [e-alfred/ocdownloader#41]
Hi there, 

On my setup I have everything containerized, even nextcloud, so Aria2 is in another container.
I would like to be able to set aria2 in another host/container, and configure it on the nextcloud settings. With docker is easy to map the same download directory on each container, but mount via nfs should be acceptable for physical environments, so download folders shouldn't be a problem. As far as I can see, the main problem is to say to ocDownloader not to search of aria2c daemon in localhost but connect via rpc to aria2 port, either the localhost or a remote host.

What do you think?

**Comment by e-alfred** [comment]
Did you try Curl as an alternative? 

Aside from thatl, to run Aria2c on a different server/container there are multiple issues that would have to be resolved:

- Authentication and setting of a custom IP would have to be implemented in ocDownloader
- The directory Aria2c downloads the files to on a separate server would have to be shared with NFS/SFTP/SAMBA and either mounted as a remote storage inside Nextcloud or the operating system.

This would have to be implemented/set up correctly to work. For now, I can think of a few quick (and rather crude) fixes to get around the first problem:

- One would be to set up either a custom internal network with Docker and change the IP here to the remote location: https://github.com/e-alfred/ocdownloader/blob/c16f7f8ae914a21613908544a8ed36ef6d9b55f8/controller/lib/aria2.php#L20

- Another option would be to use SSH port forwarding to the remote host running Aria2c and forward port 6800.

This is totally untested at the moment though.

**Comment by adocampo** [comment]
I do not knwo how to try curl instead of aria2, but anyway, it would be great to download torrents as well, and I doubt curl can download torrents.
On the other handa, from my nextcloud container, I can reach aria2 container doing a `telnet aria2_docker_ip 6800`, but I cannot reach the URL http://aria2_docker_ip:6800/jasonrpc, so, if this app works with this URL as seen on aria2.php, I doubt it will work ever... at least with this docker container. I've seen others that seems to spawn a jsonrpc, I will try with them and report back.
Thank you!

**Comment by e-alfred** [comment]
Running Curl: Change the downloader type in the admin settings and make sure that fallback.sh and fallback.php in the SERVER directory are executeable. it is described here:

https://github.com/e-alfred/ocdownloader#using-curl-instead-of-aria2

**Comment by adocampo** [comment]
It seems curl fallback mode isn't working either. I followed the steps you provided and I have both curl and curl php module installed but when I try to download (a http download) this error message appears: `Returned GID is null ! Is Aria2c running as a daemon ?`

**Comment by e-alfred** [comment]
Yes, this error usually occurs if Aria2c isn't configured properly or isn't running.

To use curl, you have to switch the downloader in the admin menu of your Nextcloud too. There is a section for ocDownloader where you can do this.

**Comment by sunjam** [comment]
Fyi, [the original repo was destroyed, but issue 18 is still available via the Wayback Machine for reference.](https://web.archive.org/web/*/https://github.com/DjazzLab/ocdownloader/issues/18)

**Comment by macinnisrr** [comment]
Love you honey, let us know how it works out :-).

**Comment by delfer** [comment]
+1 to add aria2 server url to settings

**Comment by jmlago91** [comment]
Using Aria2 in a separate container (https://github.com/P3TERX/docker-aria2-pro) and changing the host in ocdownloader/controller/lib/aria2.php:
`self::$Server = 'http://aria2:6800/jsonrpc'; `
works like a charm.

Docker compose:

`  aria2:                                                      
    image: p3terx/aria2-pro
    restart: always
    volumes:
      - /aria2/config:/config
      - /aria2/downloads:/downloads
      - /vol1/nextcloud/data:/nc_data/
    expose:
      - "6800"
    environment:
      - PUID=33
      - PGID=33
    networks:
      - back-tier
`


Nextcloud data directory shall be mounted as a volume, and after first aria2 launch, an `aria2.conf` file will appear in /aria2/config (or directory you have configured in docker-compose.yml).
There, `rpc-secret` line must be commented.
After that everything works as is.

So, I think an option in Nextcloud admin menu should be included to configure Aria2 host.

**Comment by e-alfred** [comment]
Fixed by https://github.com/e-alfred/ocdownloader/pull/164.

**Comment by jmlgo** [comment]
NC 19.0.0 and ocDownloader 1.7.8.
I don't have any configuration for address and port under Settings->Administration->Additional Settings.

Can I configure it from config.php?

**Comment by Sciroccogti** [comment]
> Nextcloud data directory shall be mounted as a volume, and after first aria2 launch, an `aria2.conf` file will appear in /aria2/config (or directory you have configured in docker-compose.yml).
> There, `rpc-secret` line must be commented.
> After that everything works as is.
> 
> So, I think an option in Nextcloud admin menu should be included to configure Aria2 host.

Hi, I've tried these lines and found out that, ocdownloader uses `/Downloads/` as default download path, so there must be a volume mounted as `/Downloads/` in aria2.

However, the downloaded files cannot be automatically added into NC's database. Should I use something like inotify or external storage?

**Comment by jmlgo** [comment]
be aware that you have to mount your nextcloud directory in aria2 in the same path as your nextcloud directory ('datadirectory') is configured in config.php from nextcloud. be sure as well that the permissions in nextcloud and aria2 containers match.
after that the app will add the files to the nextcloud database.

**Comment by chmanie** [comment]
Hey @jmlgo, I'm having troubles using your config. The connection to aria2 works but it's not reporting any progress when downloading files, it's stuck like this:

<img width="1495" alt="image" src="https://user-images.githubusercontent.com/2174084/110780494-09a44380-8265-11eb-81ee-bc1370e9f81a.png">

Also in the mounted `/downloads` directory I can't see anything. Here's my docker-compose config:

```
  aria2:
    image: p3terx/aria2-pro
    restart: always
    expose:
      - 6800
    environment:
      - PUID=82
      - PGID=82
    networks:
      - traefik_default
    volumes:
      - aria2:/config
      - nextcloud-downloads:/downloads
```

Furthermore I couldn't use the docker url in the settings:

<img width="597" alt="image" src="https://user-images.githubusercontent.com/2174084/110780948-92bb7a80-8265-11eb-9be0-009735288234.png">

I got around it by using the actual internal docker IP address.

**Comment by jmlgo** [comment]
I've mounted my nextcloud data directory in the same place for both aria2 and nextcloud containers. ocdownloader should set the destination path in aria2 for each user.
I wasn't either able to set ARIA2 Address in the admin panel. I ended up hardcoding the configuration in the ocdownloader sources.
I see that you are trying to download using HTTPS. I've only tried with bittorrent so idk if any further steps must be taken for HTTPS (I guess no).
